consensus/istanbul/backend/snapshot.go

```diff
@@ -177,6 +179,23 @@ func (s *Snapshot) apply(headers []*types.Header, gov *governance.Governance, ad
 		}
 
 		snap.ValSet, snap.Votes, snap.Tally = gov.HandleGovernanceVote(snap.ValSet, snap.Votes, snap.Tally, header, validator, addr)
+		if policy == uint64(params.WeightedRandom) {
+			// Snapshot of block N (Snapshot_N) should contain proposers for N+1 and following blocks.
+			// And proposers for Block N+1 can be calculated from the nearest previous proposersUpdateInterval block.
+			// Let's refresh proposers in Snapshot_N using previous proposersUpdateInterval block for N+1, if not updated yet.
+			pHeader := chain.GetHeaderByNumber(params.CalcProposerBlockNumber(number + 1))
+			if pHeader != nil {
+				if err := snap.ValSet.Refresh(pHeader.Hash(), pHeader.Number.Uint64(), chain.Config()); err != nil {
+					// There are three error cases and they just don't refresh proposers
+					// (1) no validator at all
+					// (2) invalid formatted hash
+					// (3) no staking info available
+					logger.Trace("Skip refreshing proposers while creating snapshot", "snap.Number", snap.Number, "pHeader.Number", pHeader.Number.Uint64(), "err", err)
+				}
+			} else {
+				logger.Trace("Can't refreshing proposers while creating snapshot due to lack of required header", "snap.Number", snap.Number)
+			}
+		}
```

Comment:
Do we need a lock for atomicity of `snap.ValSet`? @jeongkyun-oh

Reply:
I think no need to lock `snap.ValSet` which is accessed synchronously.`snap.ValSet` is an interface for `weightedCouncil`. If you worry about the `validators` and `demotedValidators`,  the `validatorsMu` protects them in `weightedCouncil`.